File: serializeTree.java
// Time Complexity : O(n) where n is the number of nodes in the tree
// Space Complexity : O(n) where n is the number of nodes in the tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
// Serialization done using BFS 
// Deserialization done using queue

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class serializeTree {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        StringBuilder sb = new StringBuilder();
        if (root == null) return sb.toString();
        q.add(root);
        while (!q.isEmpty()) {
            TreeNode curr = q.poll();
            if (curr != null) {
                sb.append(curr.val);  
                q.add(curr.left);
                q.add(curr.right);
            }   
            else sb.append("null");
            sb.append(",");   
        }
        sb.setLength(sb.length()-1);
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.isEmpty()) return null;
        String[] nodes = data.split(",");
        Queue<TreeNode> q = new LinkedList<>();
        int i = 0;
        TreeNode root = new TreeNode(Integer.parseInt(nodes[i++]));
        q.add(root);
        while (!q.isEmpty() && i < nodes.length) {
            TreeNode curr = q.poll();
            if (nodes[i].compareTo("null") != 0) {
                curr.left = new TreeNode(Integer.parseInt(nodes[i]));
                q.add(curr.left);
            }
            i++;
            if (nodes[i].compareTo("null") != 0) {
                curr.right = new TreeNode(Integer.parseInt(nodes[i]));
                q.add(curr.right);
            }
            i++;
        }
        return root;
    }
    
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));

File: serialize-tree.py
# TIME COMPLEXITY: O(N)
# SPACE COMPLEXITY: O(N)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        # Preorder traversal
        if not root:
            return []
        # Use a dequeue to store traversal - more efficiant to pop from fron while deserializing
        preorder = collections.deque()

        def helper(node):
            preorder.append(node.val)
            if node.left is not None:
                helper(node.left)
            else:  # Append null if left or right child is not present
                preorder.append(None)
            if node.right is not None:
                helper(node.right)
            else:
                preorder.append(None)
        helper(root)
        return preorder

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        # Recursively build the tree from the preorder traversal
        if len(data) == 0:
            return None

        def helper(q):
            # Base case
            if len(q) == 0:
                return None

            val = q.popleft()
            if val is None:
                return
            else:
                # Create a new treenode from popped value
                root = TreeNode(val)
                # Recursively create the left and right subtrees
                root.left = helper(q)
                root.right = helper(q)
                return root
        return helper(data)


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
